src/System.Private.CoreLib/shared/Interop/Windows/BCrypt/Interop.BCryptGenRandom.cs

```diff
@@ -10,7 +10,7 @@ internal partial class Interop
 {
     internal partial class BCrypt
     {
-        internal static unsafe int BCryptGenRandom(byte* pbBuffer, int count)
+        internal static unsafe NTSTATUS BCryptGenRandom(byte* pbBuffer, int count)
```

Comment:
We changed the one in corefx from `byte*` to `ref byte` fairly recently, so if we're converging them, the intentional change should be made here, too.

Reply:
`byte*` for buffers in PInvoke signatures is faster and more secure pattern because of the intent is clear (no need to write tricky code with `Unsafe.`, etc.).

Reply:
i remember also this https://github.com/dotnet/coreclr/pull/17528/files/c7f3d1b50bf75968c4e63376e813390688d16dbb#diff-f28b4afc50acd07a231b45a004880de5

Reply:
let me know!

Reply:
The call in CoreFX is with a Span, it was changed to `ref byte` instead of `byte*` when it got spanified.

https://github.com/dotnet/corefx/blob/3872bafe8186a3e74cb66efa4a2835d60ed67ba2/src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Windows.cs

```C#
        public override void GetBytes(Span<byte> data)
        {
            if (data.Length > 0)
            {
                GetBytes(ref MemoryMarshal.GetReference(data), data.Length);
            }
        }
...
        // Windows-specific version of this, part of the PAL.
        private static void GetBytes(ref byte pbBuffer, int count)
        {
            Debug.Assert(count > 0);

            Interop.BCrypt.NTSTATUS status = Interop.BCrypt.BCryptGenRandom(ref pbBuffer, count);
            if (status != Interop.BCrypt.NTSTATUS.STATUS_SUCCESS)
                throw Interop.BCrypt.CreateCryptographicException(status);
        }
```

Reply:
I know that folks go carried away by the new shiny thing and started using `ref` for buffer pointers in number of places. It is fine for very low level APIs where you want to optimize out the last instruction, but I do not think it is a good pattern to use for mainstream. I think your example would be better this way:

```
        public override void GetBytes(Span<byte> data)
        {
            if (data.Length > 0)
            {
                fixed (byte* p = data) GetBytes(p, data.Length);
            }
        }


        private static void GetBytes(byte* pbBuffer, int count)
        {
...
        }
```

Reply:
i wrote https://github.com/dotnet/corefx/compare/master...MarcoRossignoli:bcryptgenrandom#diff-f8fc6769feb738c3ecf772bdfe313c77R40 is implicit `&MemoryMarshal.GetReference(data)` ?

> It is fine for very low level APIs where you want to __optimize out the last instruction__

Can you add some info?

Reply:
Yes, you can pin the span just like array in latest C#.

For example, here For example, thttps://github.com/dotnet/corefx/blob/3fca26cedf6c54aabc100e69106fc7e263db7815/src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs#L82 :
```
        internal static unsafe int ToInt32(ReadOnlySpan<byte> value)
        {
            Debug.Assert(value.Length >= sizeof(int));
            return Unsafe.ReadUnaligned<int>(ref MemoryMarshal.GetReference(value));
        }
```
If you were to write this with pinning, it would be several times slower.

Reply:
updated PR, let me know if i need to change `byte*` to `ref byte`